### PR TITLE
Add Ionic 4 quickstart

### PR DESCRIPTION
**Pull request by alexisluque**
Code sample added at auth0-samples/auth0-ionic4-samples#1

### CODE REVIEW

File: articles/quickstart/native/ionic4/01-login.md [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+  path: 01-Login
+contentType: tutorial
+useCase: quickstart
+---
+
+<%= include('../_includes/_getting_started', { library: 'Ionic' }) %>
+
+<%= include('../../../_includes/_callback_url') %>
+
+The **Callback URL** to be used for your application includes your app's package ID which is found in the `config.xml` file for your app.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md:31]: If the customer hasn’t added iOS or Android as a platform, they won’t have a `config.xml` file yet.

File: articles/quickstart/native/ionic4/01-login.md [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+Go to the <a href="${manage_url}/#/applications/${account.clientId}/settings">Application Settings</a> section in your Auth0 dashboard and set your **Callback URL** in the **Allowed Callback URLs** box.
+
+If you are following along with the sample project you downloaded from the top of this page, you should set the **Allowed Callback URL** to
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: I don’t think we need to mention the sample app.  If we show them how to retrieve `YOUR_PACKAGE_ID`, they can do it in their own app or the sample app.

File: articles/quickstart/native/ionic4/01-login.md [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+```bash
+# replace YOUR_PACKAGE_ID with your app package ID
+YOUR_PACKAGE_ID://${account.namespace}/cordova/YOUR_PACKAGE_ID/callback
+```
+
+Replace `YOUR_PACKAGE_ID` with your application's package name, available as the `applicationId` attribute in the `app/build.gradle` file.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: I cannot find a `app/build.gradle` file, however just previously I was instructed to find my app’s package ID from the `config.xml` file. 

This section could probably be summed up in 3 steps
1. How to retrieve `YOUR_PACKAGE_ID`
2. How to construct the CallbackURL using YOUR_PACKAGE_ID.
3. Where to configure your CallbackURL

File: articles/quickstart/native/ionic4/01-login.md [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+
+
+<%= include('../../../_includes/_logout_url') %>
+
+::: note
+If you are following along with the sample project you downloaded from the top of this page, the logout URL you need to whitelist in the Allowed Logout URLs field is the same as the callback URL.
+:::
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: The download steps already have the user configure the logout url, so I don't think this is needed

File: articles/quickstart/native/ionic4/01-login.md [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+
+Add `file` as an allowed origin to the **Allowed Origins (CORS)** box.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: This could probably use some clarification to why the user needs to do this.

File: articles/quickstart/native/ionic4/01-login.md [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+
+```bash
+file://*
+```
+
+Lastly, be sure that the **Application Type** for your application is set to **Native** in the application settings.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: Lets link to the Application settings.

File: articles/quickstart/native/ionic4/01-login.md [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+
+```bash
+# installation with npm
+npm install auth0-js @auth0/cordova --save
+
+# installation with yarn
+yarn add auth0-js @auth0/cordova
+```
+
+Since `angular-cli` 6 and up pollyfills for node core libraries are disabled by default. As auth0-cordova uses `crypto` you'll need to provide a pollyfill for this library.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: I'm thinking maybe rewording this to

```
Auth0-Cordova requires the node library `crypto`, however since Angular CLI version 6+ you are now required to use a polyfill for this library.  You can accomplish this using a custom webpack configuration.
```

thoughts?

File: articles/quickstart/native/ionic4/01-login.md [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+// webpack.config.js
+
+module.exports = {
+	node: {
+		crypto: true
+	}
+};
+```
+
+In `angular.json` replace the `@angular-devkit/build-angular:browser` builder to `@angular-builders/custom-webpack:browser` and add customWebpackConfig to the build target options with the path to the webpack config file created prevoisly.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: `builder to` should be `builder with`

File: articles/quickstart/native/ionic4/01-login.md [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+  // ...
+}
+```
+
+### Add Cordova Plugins
+
+You must install the `SafariViewController` plugin from Cordova to be able to use universal login. The downloadable sample project already has this plugin added, but if you are embedding Auth0 in your own application, install the plugin via the command line.
+
+```bash
+ionic cordova plugin add cordova-plugin-safariviewcontroller
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md:135]: should this be 

```bash
# installation with cordova
ionic cordova plugin add cordova-plugin-safariviewcontroller
```

File: articles/quickstart/native/ionic4/01-login.md [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+```
+
+The `CustomURLScheme` plugin from Cordova is also required to handle redirects properly. The sample project has it already, but if you're adding Auth0 to your own project, install this plugin as well.
+
+```bash
+# replace YOUR_PACKAGE_ID with your app identifier
+ionic cordova plugin add cordova-plugin-customurlscheme --variable URL_SCHEME=YOUR_PACKAGE_ID --variable ANDROID_SCHEME=YOUR_PACKAGE_ID --variable ANDROID_HOST=${account.namespace} --variable ANDROID_PATHPREFIX=/cordova/YOUR_PACKAGE_ID/callback
+```
+
+Replace `YOUR_PACKAGE_ID` with your application's package name, available as the `applicationId` attribute in the `app/build.gradle` file.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: Don't have a `app/build.gradle`

File: articles/quickstart/native/ionic4/01-login.md [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+  // Needed for Auth0 (capitalization: ID):
+  clientID: '${account.clientId}',
+  // Needed for Auth0Cordova (capitalization: Id):
+  clientId: '${account.clientId}',
+  domain: '${account.namespace}',
+  packageIdentifier: 'YOUR_PACKAGE_ID' // config.xml widget ID, e.g., com.auth0.ionic
+};
+```
+
+Be sure to replace `YOUR_PACKAGE_ID` with the identifier for your app.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: This instruction might be better placed before the codeblock

File: articles/quickstart/native/ionic4/01-login.md [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+
+### Create an Authentication Service
+
+To coordinate authentication tasks, it's best to set up an injectable service that can be reused across the application. This service needs methods for logging users in and out, as well as checking their authentication state.
+
+```js
+// src/app/services/auth.service.ts
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md:256]: In our last instructions we tell the user to create a new file and where to put it, this time we use a comment at the top of the code block to show where this file should be.  Any reason for the different approaches?

File: articles/quickstart/native/ionic4/01-login.md [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+```
+
+## Add Authentication with Auth0
+
+Add a control to your app to allow users to log in. The control should call the `login` method from the `AuthService`. Start by injecting the `AuthService` in a component.
+
+```js
+// src/app/home/home.page.ts
+import { Component } from '@angular/core';
+import { AuthService } from '../services/auth.service';
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md:364]: The relative path should be `../../services/auth.service`

**Review by davidpatrick — CHANGES_REQUESTED** [review]
### Issues Starting up
When starting the app an error occurs `Cannot find module '@ionic/storage'`.  This can be resolved by installing the package, but we should instruct the user to install this if it is a new dependency.

### Issues running ionic serve
After following the instructions my app can be booted in the emulator but I run into a lot of errors with `ionic serve`

* Get blank screen and run into `Uncaught ReferenceError: global is not defined`
* After resolving that get `util.js:103 Uncaught ReferenceError: Buffer is not defined`

File: articles/quickstart/native/ionic4/01-login.md [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+    this.accessToken = null;
+    this.user = null;
+    this.loggedIn = false;
+
+    this.safariViewController.isAvailable()
+      .then((available: boolean) => {
+        const domain = AUTH_CONFIG.domain;
+        const clientId = AUTH_CONFIG.clientId;
+        const pkgId = AUTH_CONFIG.packageIdentifier;
+        let url = `https://${domain}/v2/logout?client_id=${clientId}&returnTo=${pkgId}://${domain}/cordova/${pkgId}/callback`;
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: `domain` needs to be escaped - https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#escaping-strings

File: articles/quickstart/native/ionic4/download.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+3) Set **Allowed Origins (CORS)s** in the [Application Settings](${manage_url}/#/applications/${account.clientId}/settings) to
+Android
+```text
+http://localhost, http://localhost:8100
+```
+iOS
+```text
+http://localhost, http://localhost:8100
+```
+4) Ensure that [Ionic 4](https://ionicframework.com/docs/intro/installation/), [Cordova](https://cordova.apache.org/#getstarted) and Make sure [native-run](https://github.com/ionic-team/native-run) are installed.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/download.md]: `Make` should be lowercase

File: articles/quickstart/native/ionic4/download.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+5) Check that mobile development environments for [Android](https://cordova.apache.org/docs/en/latest/guide/platforms/android/index.html) and [iOS](https://cordova.apache.org/docs/en/latest/guide/platforms/ios/index.html) are setup correctly.
+5)
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/download.md]: empty step

File: articles/quickstart/native/ionic4/01-login.md [hunk 17]
```diff
@@ -0,0 +1,443 @@
...
+
+```js
+// webpack.config.js
+
+module.exports = {
+	node: {
+		crypto: true
+	}
+};
+```
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md:102]: this code block is space indented 4 instead of 2

File: articles/quickstart/native/ionic4/01-login.md [hunk 18]
```diff
@@ -0,0 +1,443 @@
...
+ionic cordova plugin add cordova-plugin-customurlscheme --variable URL_SCHEME=YOUR_PACKAGE_ID --variable ANDROID_SCHEME=YOUR_PACKAGE_ID --variable ANDROID_HOST=${account.namespace} --variable ANDROID_PATHPREFIX=/cordova/YOUR_PACKAGE_ID/callback
+```
+
+Replace `YOUR_PACKAGE_ID` with your application's package name.
+
+## Integrate Auth0 in your Application
+
+### Modify config.xml
+
+Add `<preference name="AndroidLaunchMode" value="singleTask" />` to your config.xml. This will allow the Auth0 dialog to properly redirect back to your app.
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: this might work better as a codeblock so the user can more easily copy the code

File: articles/quickstart/native/ionic4/01-login.md [hunk 19]
```diff
@@ -0,0 +1,443 @@
...
+```
+
+## Add Authentication with Auth0
+
+Add a control to your app to allow users to log in. The control should call the `login` method from the `AuthService`. Start by injecting the `AuthService` in a component.
+
+```js
+// src/app/home/home.page.ts
+import { Component } from '@angular/core';
+import { AuthService } from '../../services/auth.service';
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: I think this is going back too far, I think it needs to be `../services/auth.service`

File: articles/quickstart/native/ionic4/01-login.md [hunk 20]
```diff
@@ -0,0 +1,443 @@
...
+        color="danger"
+        (click)="auth.logout()">Log Out</ion-button>
+    </ng-template>
+  </ng-template>
+</ion-content>
+```
+
+### Update Auth0 Dashboard Configuration
+
+1. After executing the `run` command, your `config.xml` file should contain `<allow-navigation>` tag(s).
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: I didn't have this in my `config.xml` after following all the steps

File: articles/quickstart/native/ionic4/01-login.md [hunk 21]
```diff
@@ -0,0 +1,443 @@
...
+
+```bash
+# installation with cordova
+ionic cordova plugin add cordova-plugin-safariviewcontroller
+
+# installation with npm
+npm install --save @ionic-native/safari-view-controller
+
+# installation with yarn
+yarn add @ionic-native/safari-view-controller
```
Comment (davidpatrick) [inline articles/quickstart/native/ionic4/01-login.md]: when I used yarn, it didnt actually add the plugin under cordova/plugins in my package.json

**Review by davidpatrick — COMMENTED** [review]
Besides the comments everything looks great and worked great

File: articles/quickstart/native/ionic4/01-login.md (hunk 19 above)
Comment (davidpatrick) [reply articles/quickstart/native/ionic4/01-login.md]: I think previously you had it correct with `../services/auth.service` and I had asked you to change it, not sure why...

File: articles/quickstart/native/ionic4/01-login.md (hunk 20 above)
Comment (alexisluque) [reply articles/quickstart/native/ionic4/01-login.md]: I think the section `Update Auth0 Dashboard Configuration` could be removed. This steps are only needed when changing the default config for `Ionic Web View`, but neither this tutorial and the sample are doing it.

File: articles/quickstart/native/ionic4/01-login.md (hunk 21 above)
Comment (alexisluque) [reply articles/quickstart/native/ionic4/01-login.md]: For me this is added when I run `ionic cordova plugin add cordova-plugin-safariviewcontroller`.

Comment (davidpatrick) [reply articles/quickstart/native/ionic4/01-login.md]: That might be the case, but there are 3 options to install it here.  The yarn route didn't work, at least not for me, and when viewing their official docs, they don't seem to list it as an option either https://ionicframework.com/docs/native/safari-view-controller

**Review by davidpatrick — APPROVED** [review]
